### PR TITLE
Adding BFD Enhancements

### PR DESCRIPTION
**Pull request by SumitAgarwal123**
**- What I did**

BFD Enhancements as per HLD
https://github.com/Azure/SONiC/blob/master/doc/bfd/BFD_Enhancement_HLD.md

**- How I did it**

As explained in HLD, submitting a patch

**- How to verify it**
Verified on local setup

**- Description for the changelog**

### CODE REVIEW

**Comment by SumitAgarwal123** [comment]
link to original PR
https://github.com/Azure/sonic-buildimage/pull/3385

File: dockers/docker-fpm-frr/bgpcfgd [hunk 1]
```diff
@@ -223,6 +223,61 @@ class Manager(object):
...
+        cmd = 'peer {}'.format(key_params[0])
+        if len(key_params) == 4 and key_params[3] == 'multihop':
+            cmd = cmd + ' multihop '
+        if len(key_params) > 2 and key_params[1] != 'null':
+            cmd = cmd + ' local-address ' + key_params[1]
+        if len(key_params) > 3 and key_params[2] != 'null':
+            cmd = cmd + ' interface ' + key_params[2]
+        if not data:
+            #BFD peer is deleted
+            command = "vtysh -c 'configure terminal' -c 'bfd' -c 'no {}'".format(cmd)
```
Comment (stephenxs) [inline dockers/docker-fpm-frr/bgpcfgd]: better to use list. otherwise, the run_command may fail.

File: dockers/docker-fpm-frr/bgpcfgd [hunk 2]
```diff
@@ -223,6 +223,61 @@ class Manager(object):
...
+                    command = command + " -c 'echo-mode'"
+                elif param == 'echo_interval':
+                    command = command + " -c 'echo-interval {}'".format(data[param])
+                elif param == 'label':
+                    command = command + " -c 'label {}'".format(data[param])
+                elif param == 'admin_status' and data[param] == 'up':
+                    command = command + " -c 'no shutdown'"
+                elif param == 'admin_status' and data[param] == 'down':
+                    command = command + " -c 'shutdown'"
+            run_command(command)
```
Comment (pavel-shirshov) [inline dockers/docker-fpm-frr/bgpcfgd]: Please put all this logic into a template and use vtysh -f. Please check other examples.

File: src/sonic-frr/patch/0007-BFD-Enhancement.patch [hunk 3]
```diff
@@ -0,0 +1,921 @@
...
++++ b/pimd/pim_bfd.c
+@@ -270,7 +270,7 @@ static int pim_bfd_interface_dest_update(ZAPI_CALLBACK_ARGS)
+ 			continue;
+ 		}
+ 		old_status = bfd_info->status;
+-		bfd_info->status = status;
++		BFD_SET_CLIENT_STATUS(bfd_info->status, status);
+ 		monotime(&tv);
+ 		bfd_info->last_update = tv.tv_sec;
+
```
Comment (pavel-shirshov) [inline src/sonic-frr/patch/0007-BFD-Enhancement.patch]: Why do we need this change?

Comment (SumitAgarwal123) [reply src/sonic-frr/patch/0007-BFD-Enhancement.patch]: Scenarios where this code change is required:

BFD is un-configured from BGP at remote end.
Neighbour BFD sends ADMIN_DOWN state, but BFD on local side will send
DOWN to BGP, resulting in BGP session DOWN.
Removing BFD session administratively shouldn't bring DOWN BGP session
at local or remote.

BFD is un-configured from BGP or shutdown locally.
BFD will send state DOWN to BGP resulting in BGP session DOWN.
(This is akin to saying do not use BFD for BGP)
Removing BFD session administratively shouldn't bring DOWN BGP session at
local or remote.

**Comment by SumitAgarwal123** [comment]
Abandoning this since FRR code here have moced to newer versions

**Comment by stephenxs** [comment]
Hi @SumitAgarwal123,
Will BFD be supported in 202012 or next release?

**Pull request closed without merging**